ICE with the image crate

### Code

```Rust
fn load_get<A>() -> Result<&'static A, ()> {
    unimplemented!()
}

fn main() -> Result<(), ()> {
    let image: image::ImageBuffer<image::Rgba<u8>, Vec<u8>> = load_get()?;
    Ok(())
}
```


### Meta
`rustc --version --verbose`:
```
❯ rustc --version --verbose
rustc 1.62.1 (e092d0b6b 2022-07-16)
binary: rustc
commit-hash: e092d0b6b43f2de967af0887873151bb1c0b18d3
commit-date: 2022-07-16
host: aarch64-apple-darwin
release: 1.62.1
LLVM version: 14.0.5
```

### Error output

```
error[E0275]: overflow evaluating the requirement `&num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<num_rational::Ratio<_>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>: num_traits::pow::Pow<&u128>`
 --> src/main.rs:6:63
  |
6 |     let image: image::ImageBuffer<image::Rgba<u8>, Vec<u8>> = load_get()?;
  |                                                               ^^^^^^^^^^^
  |
  = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`tester`)
```

and then if I increase the recursion limit enough...

```
process didn't exit successfully: `rustc --crate-name tester --edition=2021 src/main.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type bin --emit=dep-info,link -C embed-bitcode=no -C split-debuginfo=unpacked -C debuginfo=2 -C metadata=e9955c5db8fba40f -C extra-filename=-e9955c5db8fba40f --out-dir /Users/jjspira/Documents/Rust/tester/target/debug/deps -C incremental=/Users/jjspira/Documents/Rust/tester/target/debug/incremental -L dependency=/Users/jjspira/Documents/Rust/tester/target/debug/deps --extern image=/Users/jjspira/Documents/Rust/tester/target/debug/deps/libimage-1ae40fb2a0984fba.rlib` (signal: 10, SIGBUS: access to undefined memory)
```

adding a backtrace gives nothing new, however, compiling on nightly gives the following:

```rs
❯ cargo +nightly run
   Compiling tester v0.1.0 (/Users/jjspira/Documents/Rust/tester)

thread 'rustc' has overflowed its stack
fatal runtime error: stack overflow
error: could not compile `tester`

Caused by:
  process didn't exit successfully: `rustc --crate-name tester --edition=2021 src/main.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type bin --emit=dep-info,link -C embed-bitcode=no -C split-debuginfo=unpacked -C debuginfo=2 -C metadata=e5c31fefbe1f1d2f -C extra-filename=-e5c31fefbe1f1d2f --out-dir /Users/jjspira/Documents/Rust/tester/target/debug/deps -C incremental=/Users/jjspira/Documents/Rust/tester/target/debug/incremental -L dependency=/Users/jjspira/Documents/Rust/tester/target/debug/deps --extern image=/Users/jjspira/Documents/Rust/tester/target/debug/deps/libimage-0281d5d432728bf8.rlib` (signal: 6, SIGABRT: process abort signal)
```
</p>
</details>

I strongly suspect this is related to https://github.com/rust-lang/rust/issues/89275, but that issue is now old and should have been solved. Perhaps a regression?

--- a/compiler/rustc_typeck/src/check/coercion.rs
+++ b/compiler/rustc_typeck/src/check/coercion.rs
@@ -1479,6 +1479,10 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                 }
             }
             Err(coercion_error) => {
+                // Mark that we've failed to coerce the types here to suppress
+                // any superfluous errors we might encounter while trying to
+                // emit or provide suggestions on how to fix the initial error.
+                fcx.set_tainted_by_errors();
                 let (expected, found) = if label_expression_as_expected {
                     // In the case where this is a "forced unit", like
                     // `break`, we want to call the `()` "expected"

--- a/src/test/ui/typeck/issue-100246.rs
+++ b/src/test/ui/typeck/issue-100246.rs
@@ -1,0 +1,30 @@
+#![recursion_limit = "5"] // To reduce noise
+
+//expect incompatible type error when ambiguous traits are in scope
+//and not an overflow error on the span in the main function.
+
+struct Ratio<T>(T);
+
+pub trait Pow {
+    fn pow(self) -> Self;
+}
+
+impl<'a, T> Pow for &'a Ratio<T>
+where
+    &'a T: Pow,
+{
+    fn pow(self) -> Self {
+        self
+    }
+}
+
+fn downcast<'a, W: ?Sized>() -> std::io::Result<&'a W> {
+    todo!()
+}
+
+struct Other;
+
+fn main() -> std::io::Result<()> {
+    let other: Other = downcast()?;//~ERROR 28:24: 28:35: `?` operator has incompatible types
+    Ok(())
+}

--- a/src/test/ui/typeck/issue-100246.stderr
+++ b/src/test/ui/typeck/issue-100246.stderr
@@ -1,0 +1,13 @@
+error[E0308]: `?` operator has incompatible types
+  --> $DIR/issue-100246.rs:28:24
+   |
+LL |     let other: Other = downcast()?;
+   |                        ^^^^^^^^^^^ expected struct `Other`, found reference
+   |
+   = note: `?` operator cannot convert from `&_` to `Other`
+   = note: expected struct `Other`
+           found reference `&_`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

